cmd/workflows.go

```diff
@@ -219,50 +216,85 @@ func generateServiceAccounts(namespace *corev1.Namespace, roleBindingLister rbac
 		return nil, err
 	}
 
-	// Service account for UI access
+	// Setup the default service account
 	serviceAccounts = append(serviceAccounts, &corev1.ServiceAccount{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:        fmt.Sprintf("argo-workflows-%v", namespace.Name),
-			Namespace:   workflowsSystemNamespace,
-			Annotations: map[string]string{"workflows.argoproj.io/rbac-rule": fmt.Sprintf("'%s' in groups", roleBinding.Subjects[0].Name)},
+			Name:      "user-default-login",
+			Namespace: namespace.Name,
+			Annotations: map[string]string{
+				"workflows.argoproj.io/rbac-rule": "true",
+				"workflows.argoproj.io/rbac-rule-precedence": "0",
+			},
 		},
 	})
 
-	// Service account for workflows
+	// The service account for argo-workflows itself
 	serviceAccounts = append(serviceAccounts, &corev1.ServiceAccount{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      "argo-workflows",
 			Namespace: namespace.Name,
 		},
 	})
 
+	// The service accounts of type group used for user interface access
+	for i, subject := range roleBinding.Subjects {
+		if subject.Kind == "Group" {
+			serviceAccounts = append(serviceAccounts, &corev1.ServiceAccount{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        fmt.Sprintf("argo-workflows-%v", subject.Name),
+					Namespace:   namespace.Name,
+					Annotations: map[string]string{
+						"workflows.argoproj.io/rbac-rule": fmt.Sprintf("'%s' in groups", subject.Name),
+						"workflows.argoproj.io/rbac-rule-precedence": fmt.Sprintf("%d", i),
+					},
+				},
+			})
+		}
+	}
+
 	return serviceAccounts, nil
 }
 
 // generateRoleBindings generates role bindings for argo workflows.
-func generateRoleBindings(namespace *corev1.Namespace) []*rbacv1.RoleBinding {
+func generateRoleBindings(namespace *corev1.Namespace, roleBindingLister rbacv1listers.RoleBindingLister) ([]*rbacv1.RoleBinding, error) {
 	roleBindings := []*rbacv1.RoleBinding{}
 
-	roleBindings = append(roleBindings, &rbacv1.RoleBinding{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      fmt.Sprintf("argo-workflows-%v", namespace.Name),
-			Namespace: namespace.Name,
-		},
-		RoleRef: rbacv1.RoleRef{
-			APIGroup: rbacv1.SchemeGroupVersion.Group,
-			Kind:     "ClusterRole",
-			Name:     "argo-workflows-namespace",
-		},
-		Subjects: []rbacv1.Subject{
-			{
-				APIGroup:  "",
-				Kind:      "ServiceAccount",
-				Name:      fmt.Sprintf("argo-workflows-%v", namespace.Name),
-				Namespace: workflowsSystemNamespace,
-			},
-		},
-	})
+	// Find groups in the namespace admins
+	roleBinding, err := roleBindingLister.RoleBindings(namespace.Name).Get("namespace-admins")
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return []*rbacv1.RoleBinding{}, nil
+		}
+
+		return nil, err
+	}
 
+	// Loop over all admin groups and create a role binding for them
+	for _, subject := range roleBinding.Subjects {
+		if subject.Kind == "Group" {
```

Comment:
I would add a comment and say we bind the UI service accounts to the workflows-namespace role.